In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import numpy as  np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')


from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

In [3]:
df=pd.read_csv("/content/drive/MyDrive/DataSet/spotify.csv")
df.tail()

danceability  energy  key  loudness  mode  speechiness  acousticness  \
42300         0.528   0.693    4    -5.148     1       0.0304      0.031500   
42301         0.517   0.768    0    -7.922     0       0.0479      0.022500   
42302         0.361   0.821    8    -3.102     1       0.0505      0.026000   
42303         0.477   0.921    6    -4.777     0       0.0392      0.000551   
42304         0.529   0.945    9    -5.862     1       0.0615      0.001890   

       instrumentalness  liveness  valence  ...                      id  \
42300          0.000345    0.1210    0.394  ...  46bXU7Sgj7104ZoXxzz9tM   
42301          0.000018    0.2050    0.383  ...  0he2ViGMUO3ajKTxLOfWVT   
42302          0.000242    0.3850    0.124  ...  72DAt9Lbpy9EUS29OzQLob   
42303          0.029600    0.0575    0.488  ...  6HXgExFVuE1c3cq9QjFCcU   
42304          0.000055    0.4140    0.134  ...  6MAAMZImxcvYhRnxDLTufD   

                                        uri  \
42300  spotify:track:46bXU7Sgj7104ZoXxzz9tM   
42301  spotify:track:0he2ViGMUO3ajKTxLOfWVT   
42302  spotify:track:72DAt9Lbpy9EUS29OzQLob   
42303  spotify:track:6HXgExFVuE1c3cq9QjFCcU   
42304  spotify:track:6MAAMZImxcvYhRnxDLTufD   

                                              track_href  \
42300  https://api.spotify.com/v1/tracks/46bXU7Sgj710...   
42301  https://api.spotify.com/v1/tracks/0he2ViGMUO3a...   
42302  https://api.spotify.com/v1/tracks/72DAt9Lbpy9E...   
42303  https://api.spotify.com/v1/tracks/6HXgExFVuE1c...   
42304  https://api.spotify.com/v1/tracks/6MAAMZImxcvY...   

                                            analysis_url duration_ms  \
42300  https://api.spotify.com/v1/audio-analysis/46bX...      269208   
42301  https://api.spotify.com/v1/audio-analysis/0he2...      210112   
42302  https://api.spotify.com/v1/audio-analysis/72DA...      234823   
42303  https://api.spotify.com/v1/audio-analysis/6HXg...      323200   
42304  https://api.spotify.com/v1/audio-analysis/6MAA...      162161   

      time_signature      genre  song_name Unnamed: 0  \
42300              4  hardstyle        NaN    20995.0   
42301              4  hardstyle        NaN    20996.0   
42302              4  hardstyle        NaN    20997.0   
42303              4  hardstyle        NaN    20998.0   
42304              4  hardstyle        NaN    20999.0   

                             title  
42300           Euphoric Hardstyle  
42301  Greatest Hardstyle Playlist  
42302       Best of Hardstyle 2020  
42303           Euphoric Hardstyle  
42304       Best of Hardstyle 2020  

[5 rows x 22 columns]

##EDA

In [4]:
#how many records in given dataset
df.shape

(42305, 22)

In [5]:
df.dtypes

danceability        float64
energy              float64
key                   int64
loudness            float64
mode                  int64
speechiness         float64
acousticness        float64
instrumentalness    float64
liveness            float64
valence             float64
tempo               float64
type                 object
id                   object
uri                  object
track_href           object
analysis_url         object
duration_ms           int64
time_signature        int64
genre                object
song_name            object
Unnamed: 0          float64
title                object
dtype: object

In [6]:
#To check duplicates rows
df.duplicated().sum()

0

In [7]:
#Find the percentage of null values in each column in given dataset
df.isnull().sum()/df.shape[0]*100

danceability         0.000000
energy               0.000000
key                  0.000000
loudness             0.000000
mode                 0.000000
speechiness          0.000000
acousticness         0.000000
instrumentalness     0.000000
liveness             0.000000
valence              0.000000
tempo                0.000000
type                 0.000000
id                   0.000000
uri                  0.000000
track_href           0.000000
analysis_url         0.000000
duration_ms          0.000000
time_signature       0.000000
genre                0.000000
song_name           49.133672
Unnamed: 0          50.880511
title               50.880511
dtype: float64

In [8]:
from pandas.core.algorithms import unique
df.drop('Unnamed: 0',axis=1,inplace=True)

In [9]:
df['song_name'].mode()

0    Forever
Name: song_name, dtype: object

In [10]:
df['song_name'].fillna('NA', inplace=True)

In [11]:
df['title'].fillna('NA', inplace=True)

In [12]:
df.isnull().sum()

danceability        0
energy              0
key                 0
loudness            0
mode                0
speechiness         0
acousticness        0
instrumentalness    0
liveness            0
valence             0
tempo               0
type                0
id                  0
uri                 0
track_href          0
analysis_url        0
duration_ms         0
time_signature      0
genre               0
song_name           0
title               0
dtype: int64

In [13]:
df_cat=df.select_dtypes(object)
df_cat.dtypes

type            object
id              object
uri             object
track_href      object
analysis_url    object
genre           object
song_name       object
title           object
dtype: object

In [14]:
df_num=df.select_dtypes(["int64","float64"])
df_num.dtypes

danceability        float64
energy              float64
key                   int64
loudness            float64
mode                  int64
speechiness         float64
acousticness        float64
instrumentalness    float64
liveness            float64
valence             float64
tempo               float64
duration_ms           int64
time_signature        int64
dtype: object

In [15]:
from sklearn.preprocessing import LabelEncoder

for col in df_cat:
    le=LabelEncoder()
    df_cat[col]=le.fit_transform(df_cat[col])

In [16]:
df_new=pd.concat([df_cat,df_num],axis=1)

In [17]:
#select input and output from given dataset
X=df_new.drop('mode',axis=1)
Y=df_new['mode']

In [18]:
#train test split
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.3,
                                               random_state=1)

In [19]:
#Apply StandardScaler on input X_train and X_test
from sklearn.preprocessing import StandardScaler
ss=StandardScaler()
X_train=ss.fit_transform(X_train)#converts into numpy array
X_test=ss.transform(X_test) #converts into numpy array

##MACHINE LEARNING

In [20]:
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

#create function for train and test the model
def create_model(model):
    model.fit(X_train,Y_train)
    Y_pred=model.predict(X_test)

    #Accuracy score
    accur = accuracy_score(Y_pred,Y_test)*100

    #generate thye report
    print(classification_report(Y_test,Y_pred))
    print("Confusion Matrix:\n",confusion_matrix(Y_test,Y_pred))
    print("Accurate",accur,"% ")
    return model

In [21]:
from sklearn.linear_model import LinearRegression
lr=LogisticRegression()
lr=create_model(lr)

Y_pred = lr.predict(X_test)
lr= accuracy_score(Y_pred,Y_test)

              precision    recall  f1-score   support

           0       0.57      0.49      0.53      5684
           1       0.63      0.69      0.66      7008

    accuracy                           0.60     12692
   macro avg       0.60      0.59      0.59     12692
weighted avg       0.60      0.60      0.60     12692

Confusion Matrix:
 [[2803 2881]
 [2156 4852]]
Accurate 60.3135833595966 % 


In [22]:
#Gaussian Naive bayes
from sklearn.naive_bayes import GaussianNB
gnb=GaussianNB()
gnb=create_model(gnb)

Y_pred = gnb.predict(X_test)
gnb= accuracy_score(Y_pred,Y_test)

              precision    recall  f1-score   support

           0       0.56      0.56      0.56      5684
           1       0.64      0.64      0.64      7008

    accuracy                           0.60     12692
   macro avg       0.60      0.60      0.60     12692
weighted avg       0.60      0.60      0.60     12692

Confusion Matrix:
 [[3162 2522]
 [2533 4475]]
Accurate 60.171761739678544 % 


In [23]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier()
dt=create_model(dt)

#to create dataframe
Y_pred = dt.predict(X_test)
dt= accuracy_score(Y_pred,Y_test)

              precision    recall  f1-score   support

           0       0.65      0.67      0.66      5684
           1       0.72      0.71      0.72      7008

    accuracy                           0.69     12692
   macro avg       0.69      0.69      0.69     12692
weighted avg       0.69      0.69      0.69     12692

Confusion Matrix:
 [[3788 1896]
 [2025 4983]]
Accurate 69.10652379451622 % 


In [24]:
from sklearn.ensemble import RandomForestClassifier
rf= RandomForestClassifier()
rf=create_model(rf)

#to create dataframe
Y_pred = rf.predict(X_test)
rf= accuracy_score(Y_pred,Y_test)

              precision    recall  f1-score   support

           0       0.75      0.74      0.74      5684
           1       0.79      0.80      0.80      7008

    accuracy                           0.77     12692
   macro avg       0.77      0.77      0.77     12692
weighted avg       0.77      0.77      0.77     12692

Confusion Matrix:
 [[4183 1501]
 [1373 5635]]
Accurate 77.35581468641664 % 


In [25]:
from xgboost import XGBClassifier
xgb= XGBClassifier()
xgb=create_model(xgb)

#to create dataframe
Y_pred =xgb.predict(X_test)
xgb= accuracy_score(Y_pred,Y_test)

              precision    recall  f1-score   support

           0       0.71      0.68      0.70      5684
           1       0.75      0.77      0.76      7008

    accuracy                           0.73     12692
   macro avg       0.73      0.73      0.73     12692
weighted avg       0.73      0.73      0.73     12692

Confusion Matrix:
 [[3871 1813]
 [1583 5425]]
Accurate 73.24298770879294 % 


In [26]:
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier(n_neighbors=5,metric='minkowski',p=2)
knn= KNeighborsClassifier()
knn=create_model(knn)

#to create dataframe
Y_pred = knn.predict(X_test)
knn= accuracy_score(Y_pred,Y_test)

              precision    recall  f1-score   support

           0       0.58      0.55      0.56      5684
           1       0.65      0.68      0.67      7008

    accuracy                           0.62     12692
   macro avg       0.62      0.61      0.62     12692
weighted avg       0.62      0.62      0.62     12692

Confusion Matrix:
 [[3111 2573]
 [2228 4780]]
Accurate 62.17302237630003 % 


In [27]:
from sklearn.svm import LinearSVC
svc=LinearSVC(random_state=1)
svc= LinearSVC()
svc=create_model(svc)

#to create dataframe
Y_pred = svc.predict(X_test)
svc= accuracy_score(Y_pred,Y_test)

              precision    recall  f1-score   support

           0       0.56      0.50      0.53      5684
           1       0.63      0.69      0.66      7008

    accuracy                           0.60     12692
   macro avg       0.59      0.59      0.59     12692
weighted avg       0.60      0.60      0.60     12692

Confusion Matrix:
 [[2818 2866]
 [2190 4818]]
Accurate 60.163882760794195 % 


In [28]:
from sklearn.ensemble import AdaBoostClassifier
ada=AdaBoostClassifier(n_estimators=3,random_state=1)
ada= AdaBoostClassifier()
ada=create_model(ada)

#to create dataframe
Y_pred = ada.predict(X_test)
ada= accuracy_score(Y_pred,Y_test)

              precision    recall  f1-score   support

           0       0.68      0.64      0.66      5684
           1       0.72      0.76      0.74      7008

    accuracy                           0.71     12692
   macro avg       0.70      0.70      0.70     12692
weighted avg       0.70      0.71      0.70     12692

Confusion Matrix:
 [[3644 2040]
 [1699 5309]]
Accurate 70.54049795146548 % 


In [29]:
from sklearn.ensemble import GradientBoostingClassifier
gbc= GradientBoostingClassifier()
gbc=create_model(gbc)

#to create dataframe
Y_pred = gbc.predict(X_test)
gbc= accuracy_score(Y_pred,Y_test)

              precision    recall  f1-score   support

           0       0.69      0.65      0.67      5684
           1       0.73      0.76      0.74      7008

    accuracy                           0.71     12692
   macro avg       0.71      0.70      0.70     12692
weighted avg       0.71      0.71      0.71     12692

Confusion Matrix:
 [[3681 2003]
 [1685 5323]]
Accurate 70.94232587456666 % 


In [30]:
score = [lr,gnb,dt,rf,xgb,knn,svc,ada,gbc]
Models = pd.DataFrame({'n_neighbors': ["Logistic Regression","Gaussian Naive Bayes","Decision Tree","Random Forest","XGBoost", "K-Neighbors","Support vector","ADA BOOST","Gradient Boost"],'Accuracy': score})
Models.sort_values(by='Score',ascending=False)

n_neighbors     Score
3         Random Forest  0.773558
4               XGBoost  0.732430
8        Gradient Boost  0.709423
7             ADA BOOST  0.705405
2         Decision Tree  0.691065
5           K-Neighbors  0.621730
0   Logistic Regression  0.603136
1  Gaussian Naive Bayes  0.601718
6        Support vector  0.601639

In [31]:
#CONCLUSION
'''
from the above list we get to know that for this Spotify dataset RandomForest
alogorithm is best with the result of the 77.35% .
'''

'\nfrom the above list we get to know that for this Spotify dataset RandomForest\nalogorithm is best with the result of the 77.32% .\n'

##DEEP LEARNING

In [32]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.models import load_model

In [52]:
from tensorflow.keras.callbacks import EarlyStopping

model=Sequential()

model.add(Dense(64, input_dim=20, activation="relu"))
model.add(Dropout(rate=0.05))
model.add(Dense(32, activation="relu"))
model.add(Dropout(rate=0.05))
model.add(Dense(16, activation="relu"))
model.add(Dropout(rate=0.05))
model.add(Dense(8, activation="relu"))
model.add(Dropout(rate=0.05))
model.add(Dense(1, activation="sigmoid"))

model.compile(optimizer="adam",loss="binary_crossentropy", metrics=["accuracy"]) #complile model
earlystop=EarlyStopping(monitor="val_loss",mode="min",verbose=1,patience=50) #use early stopping

model.fit(X_train,Y_train,epochs=500,validation_data=(X_test,Y_test),callbacks=[earlystop], batch_size=16) #train data


Epoch 1/500
1851/1851 [==============================] - 5s 2ms/step - loss: 0.6550 - accuracy: 0.5943 - val_loss: 0.6345 - val_accuracy: 0.6233
Epoch 2/500
1851/1851 [==============================] - 4s 2ms/step - loss: 0.6379 - accuracy: 0.6212 - val_loss: 0.6299 - val_accuracy: 0.6517
Epoch 3/500
1851/1851 [==============================] - 4s 2ms/step - loss: 0.6274 - accuracy: 0.6538 - val_loss: 0.6146 - val_accuracy: 0.6740
Epoch 4/500
1851/1851 [==============================] - 4s 2ms/step - loss: 0.6167 - accuracy: 0.6741 - val_loss: 0.6080 - val_accuracy: 0.6867
Epoch 5/500
1851/1851 [==============================] - 4s 2ms/step - loss: 0.6097 - accuracy: 0.6806 - val_loss: 0.5985 - val_accuracy: 0.6913
Epoch 6/500
1851/1851 [==============================] - 4s 2ms/step - loss: 0.6041 - accuracy: 0.6867 - val_loss: 0.5966 - val_accuracy: 0.6900
Epoch 7/500
1851/1851 [==============================] - 4s 2ms/step - loss: 0.6002 - accuracy: 0.6900 - val_loss: 0.5954 - val_ac

In [53]:
# Generate predictions
y_pred = model.predict(X_test)
y_pred = np.round(y_pred).astype(int)

397/397 [==============================] - 0s 858us/step


In [54]:
# Print classification report
print(classification_report(Y_test, y_pred))

              precision    recall  f1-score   support

           0       0.66      0.68      0.67      5684
           1       0.74      0.71      0.72      7008

    accuracy                           0.70     12692
   macro avg       0.70      0.70      0.70     12692
weighted avg       0.70      0.70      0.70     12692



In [44]:
# Calculate accuracy
accuracy = (Y_test.values == y_pred).mean() * 100
print(f"Accuracy: {accuracy}")

Accuracy: 50.71013611637507


##CONCLUSION

##By comparing ML and DL from the above, we get to know that for this Spotify dataset
ML RandomForest alogorithm is best with the result of the 77.32% accuracy.
